File: frontend/src/services/AccountingEngine.test.ts
import { describe, it, expect } from 'vitest';
import { AccountingEngine } from './AccountingEngine';
import { DoublingLevel, ProjectType } from '../types';

describe('AccountingEngine', () => {

    // ═══════════════════════════════════════════════════════════
    // SUN Scoring
    // ═══════════════════════════════════════════════════════════
    describe('SUN scoring', () => {
        it('should convert abnat to game points using (abnat × 2) / 10', () => {
            // 65 abnat → (65*2)/10 = 13 game points each
            const result = AccountingEngine.calculateRoundResult(
                65, 65, 0, 0, 'SUN', 1 as DoublingLevel, null
            );
            expect(result.us.gamePoints).toBe(13);
            expect(result.them.gamePoints).toBe(13);
        });

        it('should handle normal SUN round with bidder winning', () => {
            // Us bid SUN, scored 80 abnat (80*2/10=16 > 13 → win)
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.us.gamePoints).toBe(16);
            expect(result.them.gamePoints).toBe(10);
            expect(result.winner).toBe('us');
        });

        it('should handle odd abnat creating .X in SUN', () => {
            // 43 abnat → (43*2)/10 = 8.6
            const result = AccountingEngine.calculateRoundResult(
                43, 87, 0, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            // 43*2/10 = 8.6, 87*2/10 = 17.4
            // bidder(us) scored 8.6 ≤ 13 → Khasara
            expect(result.us.gamePoints).toBe(0);
            expect(result.them.gamePoints).toBe(26);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // HOKUM Scoring
    // ═══════════════════════════════════════════════════════════
    describe('HOKUM scoring', () => {
        it('should round 0.5 DOWN in HOKUM conversion', () => {
            // 85 abnat → 85/10 = 8.5 → floors to 8
            // 77 abnat → 77/10 = 7.7 → ceil to 8
            // Sum = 16 = maxPoints → exact fit
            const result = AccountingEngine.calculateRoundResult(
                85, 77, 0, 0, 'HOKUM', 1 as DoublingLevel, null
            );
            expect(result.us.gamePoints).toBe(8);
            expect(result.them.gamePoints).toBe(8);
        });

        it('should handle sum > maxPoints by capping loser', () => {
            // 86/10=8.6→9, 76/10=7.6→8. Sum=17>16. Winner(us, 9>8) keeps 9, them=16-9=7
            const result = AccountingEngine.calculateRoundResult(
                86, 76, 0, 0, 'HOKUM', 1 as DoublingLevel, null
            );
            expect(result.us.gamePoints).toBe(9);
            expect(result.them.gamePoints).toBe(7);
        });

        it('should handle sum < maxPoints by adding remainder to winner', () => {
            // 84/10=8.4→8, 75/10=7.5→7. Sum=15<16. Remainder=1 to winner(us)→9
            const result = AccountingEngine.calculateRoundResult(
                84, 75, 0, 0, 'HOKUM', 1 as DoublingLevel, null
            );
            expect(result.us.gamePoints).toBe(9);
            expect(result.them.gamePoints).toBe(7);
        });

        it('should handle HOKUM max raw (162 abnat → 16 pts)', () => {
            // 162/10=16.2→16. themRaw=0 → Kaboot
            const result = AccountingEngine.calculateRoundResult(
                162, 0, 0, 0, 'HOKUM', 1 as DoublingLevel, 'us'
            );
            expect(result.us.isKaboot).toBe(true);
            expect(result.us.gamePoints).toBe(25);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // Kaboot (Capot - All Tricks Won)
    // ═══════════════════════════════════════════════════════════
    describe('Kaboot', () => {
        it('should award 44 points for SUN Kaboot', () => {
            const result = AccountingEngine.calculateRoundResult(
                130, 0, 0, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.us.isKaboot).toBe(true);
            expect(result.us.gamePoints).toBe(44);
            expect(result.them.gamePoints).toBe(0);
        });

        it('should award 25 points for HOKUM Kaboot', () => {
            const result = AccountingEngine.calculateRoundResult(
                162, 0, 0, 0, 'HOKUM', 1 as DoublingLevel, 'us'
            );
            expect(result.us.isKaboot).toBe(true);
            expect(result.us.gamePoints).toBe(25);
            expect(result.them.gamePoints).toBe(0);
        });

        it('should detect Kaboot for opponent team', () => {
            const result = AccountingEngine.calculateRoundResult(
                0, 162, 0, 0, 'HOKUM', 1 as DoublingLevel, 'them'
            );
            expect(result.them.isKaboot).toBe(true);
            expect(result.them.gamePoints).toBe(25);
            expect(result.us.gamePoints).toBe(0);
        });

        it('should NOT detect Kaboot if both are 0', () => {
            const result = AccountingEngine.calculateRoundResult(
                0, 0, 0, 0, 'SUN', 1 as DoublingLevel, null
            );
            expect(result.us.isKaboot).toBe(false);
            expect(result.them.isKaboot).toBe(false);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // Khasara (Buyer Loss)
    // ═══════════════════════════════════════════════════════════
    describe('Khasara', () => {
        it('should give opponent ALL 26 points in SUN if bidder scores ≤ 13', () => {
            // Bidder (us) scored exactly 65 abnat → 13 pts → ≤ 13 → KHASARA
            const result = AccountingEngine.calculateRoundResult(
                65, 65, 0, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.us.gamePoints).toBe(0);
            expect(result.them.gamePoints).toBe(26);
        });

        it('should give opponent ALL 16 points in HOKUM if bidder scores ≤ 8', () => {
            // Bidder (us) scored 80 abnat → 80/10=8 → ≤ 8 → KHASARA
            const result = AccountingEngine.calculateRoundResult(
                80, 82, 0, 0, 'HOKUM', 1 as DoublingLevel, 'us'
            );
            expect(result.us.gamePoints).toBe(0);
            expect(result.them.gamePoints).toBe(16);
        });

        it('should NOT trigger Khasara if bidder scores > threshold', () => {
            // Bidder (us) scored 70 abnat → 70*2/10=14 pts in SUN → > 13 → WIN
            const result = AccountingEngine.calculateRoundResult(
                70, 60, 0, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.us.gamePoints).toBe(14);
            expect(result.them.gamePoints).toBe(12);
        });

        it('should handle Khasara when opponent is bidder', () => {
            // Them bid SUN, them scored 60 abnat → 12 pts → ≤ 13 → KHASARA
            const result = AccountingEngine.calculateRoundResult(
                70, 60, 0, 0, 'SUN', 1 as DoublingLevel, 'them'
            );
            expect(result.them.gamePoints).toBe(0);
            expect(result.us.gamePoints).toBe(26);
        });

        it('should NOT trigger Khasara if bidderTeam is null', () => {
            // No bidder specified → no Khasara check
            const result = AccountingEngine.calculateRoundResult(
                65, 65, 0, 0, 'SUN', 1 as DoublingLevel, null
            );
            expect(result.us.gamePoints).toBe(13);
            expect(result.them.gamePoints).toBe(13);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // Doubling
    // ═══════════════════════════════════════════════════════════
    describe('Doubling', () => {
        it('should multiply total and give all to winner when doubled', () => {
            // SUN: Us=80(16pts), Them=50(10pts). Total=26. Winner(us): 26×2=52
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 2 as DoublingLevel, 'us'
            );
            expect(result.us.gamePoints).toBe(52);
            expect(result.them.gamePoints).toBe(0);
        });

        it('should handle tripling (×3)', () => {
            // SUN: Us=80(16pts), Them=50(10pts). Total=26. Winner(us): 26×3=78
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 3 as DoublingLevel, 'us'
            );
            expect(result.us.gamePoints).toBe(78);
            expect(result.them.gamePoints).toBe(0);
        });

        it('should not multiply if level is 1', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.us.gamePoints).toBe(16);
            expect(result.them.gamePoints).toBe(10);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // Baloot Bonus
    // ═══════════════════════════════════════════════════════════
    describe('Baloot bonus', () => {
        it('should add flat 2 points for Baloot (never multiplied)', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 1 as DoublingLevel, 'us', true, false
            );
            expect(result.us.gamePoints).toBe(18); // 16 + 2
            expect(result.them.gamePoints).toBe(10);
        });

        it('should add Baloot for both teams independently', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 1 as DoublingLevel, 'us', true, true
            );
            expect(result.us.gamePoints).toBe(18);  // 16 + 2
            expect(result.them.gamePoints).toBe(12); // 10 + 2
        });

        it('should NOT multiply Baloot bonus even when doubled', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 2 as DoublingLevel, 'us', true, false
            );
            // Winner (us): 26 × 2 = 52 + 2 Baloot = 54
            expect(result.us.gamePoints).toBe(54);
            expect(result.them.gamePoints).toBe(0);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // Project Abnat Values
    // ═══════════════════════════════════════════════════════════
    describe('getProjectAbnatValue', () => {
        it('should return correct values for FOUR_HUNDRED', () => {
            expect(AccountingEngine.getProjectAbnatValue(ProjectType.FOUR_HUNDRED, 'SUN')).toBeGreaterThan(0);
            expect(AccountingEngine.getProjectAbnatValue(ProjectType.FOUR_HUNDRED, 'HOKUM')).toBe(0);
        });

        it('should return values for HUNDRED in SUN and HOKUM', () => {
            const sunVal = AccountingEngine.getProjectAbnatValue(ProjectType.HUNDRED, 'SUN');
            const hokumVal = AccountingEngine.getProjectAbnatValue(ProjectType.HUNDRED, 'HOKUM');
            expect(sunVal).toBeGreaterThan(hokumVal);
            expect(hokumVal).toBeGreaterThan(0);
        });

        it('should return values for FIFTY', () => {
            const sunVal = AccountingEngine.getProjectAbnatValue(ProjectType.FIFTY, 'SUN');
            const hokumVal = AccountingEngine.getProjectAbnatValue(ProjectType.FIFTY, 'HOKUM');
            expect(sunVal).toBeGreaterThan(0);
            expect(hokumVal).toBeGreaterThan(0);
        });

        it('should return values for SIRA', () => {
            const sunVal = AccountingEngine.getProjectAbnatValue(ProjectType.SIRA, 'SUN');
            const hokumVal = AccountingEngine.getProjectAbnatValue(ProjectType.SIRA, 'HOKUM');
            expect(sunVal).toBeGreaterThan(0);
            expect(hokumVal).toBeGreaterThan(0);
        });

        it('should return 0 for BALOOT (handled as flat game points)', () => {
            expect(AccountingEngine.getProjectAbnatValue(ProjectType.BALOOT, 'SUN')).toBe(0);
            expect(AccountingEngine.getProjectAbnatValue(ProjectType.BALOOT, 'HOKUM')).toBe(0);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // Projects in Scoring
    // ═══════════════════════════════════════════════════════════
    describe('Projects in scoring', () => {
        it('should include project abnat in totalRaw', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 20, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.us.totalRaw).toBe(100); // 80 + 20
            // Game points: 100 * 2 / 10 = 20
            expect(result.us.gamePoints).toBe(20);
        });

        it('should include opponent projects in their totalRaw', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 10, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.them.totalRaw).toBe(60); // 50 + 10
            expect(result.them.gamePoints).toBe(12); // 60*2/10
        });
    });

    // ═══════════════════════════════════════════════════════════
    // ScoreBreakdown Structure
    // ═══════════════════════════════════════════════════════════
    describe('ScoreBreakdown structure', () => {
        it('should populate all breakdown fields', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 10, 5, 'SUN', 1 as DoublingLevel, 'us'
            );

            expect(result.us.rawCardPoints).toBe(80);
            expect(result.us.projectPoints).toBe(10);
            expect(result.us.totalRaw).toBe(90);
            expect(result.us.isKaboot).toBe(false);
            expect(result.us.multiplierApplied).toBe(1);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // Edge Cases
    // ═══════════════════════════════════════════════════════════
    describe('Edge cases', () => {
        it('should handle tie correctly (no bidder)', () => {
            const result = AccountingEngine.calculateRoundResult(
                65, 65, 0, 0, 'SUN', 1 as DoublingLevel, null
            );
            expect(result.winner).toBe('tie');
        });

        it('should handle zero-sum game', () => {
            const result = AccountingEngine.calculateRoundResult(
                0, 0, 0, 0, 'SUN', 1 as DoublingLevel, null
            );
            expect(result.us.gamePoints).toBe(0);
            expect(result.them.gamePoints).toBe(0);
        });

        it('should return baida as false (placeholder)', () => {
            const result = AccountingEngine.calculateRoundResult(
                80, 50, 0, 0, 'SUN', 1 as DoublingLevel, 'us'
            );
            expect(result.baida).toBe(false);
        });
    });

    // ═══════════════════════════════════════════════════════════
    // explainCalculation
    // ═══════════════════════════════════════════════════════════
    describe('explainCalculation', () => {
        it('should return non-empty array of explanation steps', () => {
            const steps = AccountingEngine.explainCalculation(80, 50, 'SUN', 'us');
            expect(steps.length).toBeGreaterThan(0);
        });

        it('should include bid type in first step', () => {
            const steps = AccountingEngine.explainCalculation(80, 50, 'SUN', 'us');
            expect(steps[0]).toContain('SUN');
        });

        it('should work for HOKUM as well', () => {
            const steps = AccountingEngine.explainCalculation(90, 72, 'HOKUM', 'them');
            expect(steps.length).toBeGreaterThan(0);
            expect(steps[0]).toContain('HOKUM');
        });
    });
});

File: mobile/test/services/accounting_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:baloot_ai/services/accounting_engine.dart';
import 'package:baloot_ai/models/enums.dart';

void main() {
  group('AccountingEngine — SUN Conversion', () {
    test('sunCardGP: floor-to-even rounding', () {
      // q odd + remainder > 0 → round up
      expect(AccountingEngine.sunCardGP(67), 14); // 13 r 2, odd → 14
      expect(AccountingEngine.sunCardGP(63), 12); // 12 r 3, even → 12
      // Edge cases
      expect(AccountingEngine.sunCardGP(0), 0);
      expect(AccountingEngine.sunCardGP(5), 1); // 1 r 0, no remainder
      expect(AccountingEngine.sunCardGP(130), 26); // Full pool
    });

    test('sunCardGP: pairs always sum to 26', () {
      // Test several complementary pairs that sum to 130
      final pairs = [
        [65, 65],
        [70, 60],
        [80, 50],
        [90, 40],
        [100, 30],
        [110, 20],
        [120, 10],
        [130, 0],
        [67, 63],
        [73, 57],
      ];
      for (final pair in pairs) {
        final gpA = AccountingEngine.sunCardGP(pair[0]);
        final gpB = AccountingEngine.sunCardGP(pair[1]);
        expect(gpA + gpB, 26,
            reason: 'SUN ${pair[0]}+${pair[1]} should sum to 26, got $gpA+$gpB');
      }
    });
  });

  group('AccountingEngine — HOKUM Conversion', () {
    test('hokumCardGP: individual rounding (r > 5 rounds up)', () {
      expect(AccountingEngine.hokumCardGP(85), 8); // 8 r 5 → 8 (r=5 is NOT > 5)
      expect(AccountingEngine.hokumCardGP(86), 9); // 8 r 6 → 9
      expect(AccountingEngine.hokumCardGP(80), 8); // 8 r 0 → 8
      expect(AccountingEngine.hokumCardGP(0), 0);
      expect(AccountingEngine.hokumCardGP(162), 16); // Full pool
    });

    test('hokumPairGP: pairs always sum to 16', () {
      final pairs = [
        [81, 81],
        [90, 72],
        [100, 62],
        [110, 52],
        [120, 42],
        [130, 32],
        [150, 12],
        [162, 0],
        [85, 77],
        [96, 66],
      ];
      for (final pair in pairs) {
        final result = AccountingEngine.hokumPairGP(pair[0], pair[1]);
        expect(result[0] + result[1], 16,
            reason:
                'HOKUM ${pair[0]}+${pair[1]} should sum to 16, got ${result[0]}+${result[1]}');
      }
    });

    test('hokumPairGP: sum=17 reduces larger remainder side', () {
      // Both individually round up, but sum exceeds 16
      final result = AccountingEngine.hokumPairGP(86, 76);
      expect(result[0] + result[1], 16);
    });

    test('hokumPairGP: sum=15 increases larger remainder side', () {
      // Both individually round down, but sum below 16
      final result = AccountingEngine.hokumPairGP(85, 77);
      expect(result[0] + result[1], 16);
    });
  });

  group('AccountingEngine — Kaboot (Capot)', () {
    test('SUN kaboot: winner gets 44, loser gets 0', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 130,
        themRaw: 0,
        usProjects: 0,
        themProjects: 0,
        bidType: 'SUN',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'us',
      );
      expect(result.us.gamePoints, 44);
      expect(result.them.gamePoints, 0);
      expect(result.us.isKaboot, true);
      expect(result.winner, 'us');
    });

    test('HOKUM kaboot: winner gets 25, loser gets 0', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 0,
        themRaw: 162,
        usProjects: 0,
        themProjects: 0,
        bidType: 'HOKUM',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'them',
      );
      expect(result.them.gamePoints, 25);
      expect(result.us.gamePoints, 0);
      expect(result.them.isKaboot, true);
      expect(result.winner, 'them');
    });
  });

  group('AccountingEngine — Khasara (Buyer Loss)', () {
    test('bidder GP < opponent GP → khasara', () {
      // Us bid SUN, scored less
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 40,
        themRaw: 90,
        usProjects: 0,
        themProjects: 0,
        bidType: 'SUN',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'us',
      );
      // Us scored 8 GP, Them scored 18 GP → khasara → them get 26
      expect(result.us.gamePoints, 0);
      expect(result.them.gamePoints, 26);
      expect(result.winner, 'them');
    });

    test('GP tie with bidder raw < opp raw → khasara', () {
      // Carefully picked so GP are equal but raw abnat differ
      // SUN: 33 → 6 r 3, even → 6 GP. 97 → 19 r 2, odd → 20. No, sum != 26
      // Let's use HOKUM with projects to create a tie
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 80,
        themRaw: 82,
        usProjects: 0,
        themProjects: 0,
        bidType: 'HOKUM',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'us',
      );
      // HOKUM pair: 80→8, 82→8+1=9? No: 80/10=8r0→8, 82/10=8r2→8. Sum=16. OK.
      // Wait: 80+82=162, pair should sum to 16
      // 80: q=8,r=0→8. 82: q=8,r=2→8. Sum=16. ✓
      // Us GP=8, Them GP=8 → tie → compare raw: 80 < 82 → khasara
      expect(result.us.gamePoints, 0);
      expect(result.them.gamePoints, 16);
    });

    test('GP tie, doubled → doubler always loses', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 81,
        themRaw: 81,
        usProjects: 0,
        themProjects: 0,
        bidType: 'HOKUM',
        doublingLevel: DoublingLevel.double_,
        bidderTeam: 'us',
      );
      // 81: q=8,r=1→8 each. Sum=16. ✓
      // Tie + doubled → bidder (us) loses
      // Khasara: them get 16 base. Then doubling: 16*2=32 to them
      expect(result.us.gamePoints, 0);
      expect(result.them.gamePoints, 32);
    });

    test('GP tie, equal raw → split (no khasara)', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 81,
        themRaw: 81,
        usProjects: 0,
        themProjects: 0,
        bidType: 'HOKUM',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'us',
      );
      // 81→8 each, sum=16. ✓. Raw equal → no khasara → split
      expect(result.us.gamePoints, 8);
      expect(result.them.gamePoints, 8);
    });
  });

  group('AccountingEngine — Baloot Bonus', () {
    test('baloot adds flat 2 GP, never multiplied', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 90,
        themRaw: 40,
        usProjects: 0,
        themProjects: 0,
        bidType: 'SUN',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'us',
        hasBalootUs: true,
      );
      // 90→18, 40→8. Us wins. +2 baloot = 20
      expect(result.us.gamePoints, 20);
      expect(result.them.gamePoints, 8);
    });

    test('baloot survives khasara for winning team', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 40,
        themRaw: 90,
        usProjects: 0,
        themProjects: 0,
        bidType: 'SUN',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'us',
        hasBalootThem: true,
      );
      // Us bid, scored 8 GP < 18 GP → khasara → them get 26 + 2 baloot = 28
      expect(result.them.gamePoints, 28);
      expect(result.us.gamePoints, 0);
    });

    test('baloot not multiplied even when doubled', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 100,
        themRaw: 62,
        usProjects: 0,
        themProjects: 0,
        bidType: 'HOKUM',
        doublingLevel: DoublingLevel.double_,
        bidderTeam: 'us',
        hasBalootUs: true,
      );
      // HOKUM pair: 100→10, 62→6. Sum=16. ✓
      // Us wins. Doubled: total=16, us gets 16*2=32. Then +2 baloot = 34.
      expect(result.us.gamePoints, 34);
      expect(result.them.gamePoints, 0);
    });
  });

  group('AccountingEngine — Doubling', () {
    test('double multiplier: winner takes all × 2', () {
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 100,
        themRaw: 62,
        usProjects: 0,
        themProjects: 0,
        bidType: 'HOKUM',
        doublingLevel: DoublingLevel.double_,
        bidderTeam: 'us',
      );
      // HOKUM pair: 100→10, 62→6. Sum=16.
      // Us wins: total=16, doubled=16*2=32
      expect(result.us.gamePoints, 32);
      expect(result.them.gamePoints, 0);
    });
  });

  group('AccountingEngine — Project Abnat Values', () {
    test('SUN project values', () {
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.fourHundred, 'SUN'), 40);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.hundred, 'SUN'), 20);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.fifty, 'SUN'), 10);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.sira, 'SUN'), 4);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.baloot, 'SUN'), 0);
    });

    test('HOKUM project values', () {
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.fourHundred, 'HOKUM'), 0);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.hundred, 'HOKUM'), 10);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.fifty, 'HOKUM'), 5);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.sira, 'HOKUM'), 2);
      expect(AccountingEngine.getProjectAbnatValue(ProjectType.baloot, 'HOKUM'), 0);
    });
  });

  group('AccountingEngine — Projects in Scoring', () {
    test('projects add to total raw before conversion', () {
      // SUN: usRaw=60 + usProjects=20 = 80 total
      //       themRaw=70 + themProjects=0 = 70 total
      final result = AccountingEngine.calculateRoundResult(
        usRaw: 60,
        themRaw: 70,
        usProjects: 20,
        themProjects: 0,
        bidType: 'SUN',
        doublingLevel: DoublingLevel.normal,
        bidderTeam: 'us',
      );
      // SUN: 80→16, 70 (wait: 60+70=130, projects add on top)
      // Us totalRaw=80, them totalRaw=70. 80+70=150 (not 130, projects add)
      // sunCardGP(80) = 80/5=16 r 0 → 16 (even, no round)
      // sunCardGP(70) = 70/5=14 r 0 → 14
      // Us: 16, them: 14 → us wins
      expect(result.us.gamePoints, 16);
      expect(result.them.gamePoints, 14);
    });
  });

  group('AccountingEngine — Explain Calculation', () {
    test('returns step-by-step explanation for SUN', () {
      final steps = AccountingEngine.explainCalculation(
        usRaw: 70,
        themRaw: 60,
        bidType: 'SUN',
        bidderTeam: 'us',
      );
      expect(steps.length, greaterThan(3));
      expect(steps[0], contains('SUN'));
      expect(steps.any((s) => s.contains('Us:')), true);
    });

    test('returns step-by-step explanation for HOKUM', () {
      final steps = AccountingEngine.explainCalculation(
        usRaw: 100,
        themRaw: 62,
        bidType: 'HOKUM',
        bidderTeam: 'them',
      );
      expect(steps.length, greaterThan(3));
      expect(steps[0], contains('HOKUM'));
    });
  });
}
